src/gleam/list.gleam

```diff
@@ -454,6 +454,64 @@ pub fn fold_right(list: List(a), from initial: b, with fun: fn(a, b) -> b) -> b
   }
 }
 
+/// Like fold but the folding function also receives the index of the current element.
+///
+/// ## Examples
+///
+/// ```
+/// ["a", "b", "c"]
+/// |> list.index_fold([], fn(index, item, acc) { ... })
+/// ```
+///
+pub fn index_fold(
+  over over: List(a),
+  from initial: b,
+  with fun: fn(Int, a, b) -> b,
+) -> b {
+  over
+  |> index_map(fn(ix, value) { tuple(ix, value) })
+  |> fold(
+    from: initial,
+    with: fn(t, acc) {
+      let tuple(ix, val) = t
+      fun(ix, val, acc)
+    },
+  )
+}
+
+/// A variant of fold that allows to stop folding earlier.
+///
+/// The folding function should return `Result(accumulator, accumulator)
+/// If the returned value is `Ok(accumulator)` try_fold will try the next value in the list.
+/// If the returned value is `Error(accumulator)` try_fold will stop and return that accumulator.
+///
+/// ## Examples
+///
+/// ```
+/// [1, 2, 3, 4]
+/// |> try_fold(0, fn(i, acc) {
+///   case i < 3 {
+///     True -> Ok(acc + i)
+///     False -> Error(acc)
+///   }
+/// })
+/// ```
+///
+pub fn try_fold(
```

Comment:
I think this should have the sig `fn(List(a), b, fn(a, b) -> Result(b, e)) -> Result(b, e)` to match the other `try` functions.

Reply:
Oh, I see, it will return the error.
I had a different use case in mind, will change that and propose a new function for what I wanted.

Reply:
This is what I had in mind before https://ocaml.janestreet.com/ocaml-core/v0.13/doc/base/Base/List/index.html#val-fold_until